---
title: "PHDS Data Visualization Workshop: Single Cell and Spatial Data"
author: "Lukas M. Weber"
date: "2024-03-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction

In this session we will use example datasets from single-cell RNA sequencing and spatial transcriptomics.

We will use R packages from [Bioconductor](https://bioconductor.org/) to process and analyze the datasets.


## Load packages

Note that CRAN packages can be installed with `install.packages()`, and Bioconductor packages can be installed with `BiocManager::install()` (after first installing `BiocManager` with `install.packages("BiocManager")`).

```{r, message=FALSE}
# load Bioconductor packages
library(SingleCellExperiment)
library(scater)
library(scran)
library(SingleR)
```


# Single-cell RNA sequencing dataset



# Spatial transcriptomics dataset

## Load packages

```{r, message=FALSE}
# load Bioconductor packages
library(SpatialExperiment)
library(STexampleData)
library(scater)
library(scran)
```


## Load data

```{r}
# load dataset from STexampleData package
spe <- Visium_humanDLPFC()
```


## Analysis workflow

```{r}
# preprocessing

# subset to keep only spots over tissue
spe <- spe[, colData(spe)$in_tissue == 1]

# identify mitochondrial genes
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe)$gene_name)

# calculate per-spot QC metrics
spe <- addPerCellQC(spe, subsets = list(mito = is_mito))

# no additional quality control needed for this dataset
```

```{r}
# calculate logcounts
spe <- logNormCounts(spe)
```

```{r}
# feature selection

# remove mitochondrial genes
spe <- spe[!is_mito, ]
# fit mean-variance relationship
dec <- modelGeneVar(spe)
# select top HVGs
top_hvgs <- getTopHVGs(dec, prop = 0.1)
```

```{r}
# dimensionality reduction

# compute PCA
set.seed(123)
spe <- runPCA(spe, subset_row = top_hvgs)

# compute UMAP on top 50 PCs
set.seed(123)
spe <- runUMAP(spe, dimred = "PCA")
# update column names
colnames(reducedDim(spe, "UMAP")) <- paste0("UMAP", 1:2)
```

```{r}
# clustering

# graph-based clustering
set.seed(123)
k <- 10
g <- buildSNNGraph(spe, k = k, use.dimred = "PCA")
g_walk <- igraph::cluster_walktrap(g)
clus <- g_walk$membership
colLabels(spe) <- factor(clus)
```


# Visualizations

We use functions available in the `ggspavis` package, and also demonstrate how to create these using `ggplot2` code.

```{r, message=FALSE}
library(ggspavis)
```


## Spatial plots

Here we demonstrate how to generate spatial plots in x-y coordinates.

Basic x-y plot:

```{r}
plotSpots(spe)
```

Annotate by color using a categorical variable (reference cluster labels):

```{r}
plotSpots(spe, annotate = "ground_truth")
```

Adjust color palette:

```{r}
plotSpots(spe, annotate = "ground_truth", pal = "libd_layer_colors")
```

Annotate by color using a different categorical variable (clustering labels):

```{r}
plotSpots(spe, annotate = "label", pal = "libd_layer_colors")
```

Annotate by color using continuous variable (expression of a specific gene):

```{r}
plotSpots(spe, annotate = "PCP4")
```


## Reduced dimension plots

Principal component analysis (PCA):

```{r}
plotDimRed(spe, plot_type = "PCA")
```

Principal component analysis (PCA) with annotation by a categorical variable:

```{r}
plotDimRed(spe, plot_type = "PCA", annotate = "label")
```

UMAP:

```{r}
plotDimRed(spe, plot_type = "UMAP")
```

UMAP with annotation by a categorical variable:

```{r}
plotDimRed(spe, plot_type = "UMAP", annotate = "label")
```
